discojs/src/aggregator/byzantine.ts

```diff
@@ -0,0 +1,185 @@
+                debug(
+                    "[WARN] Setting the aggregator's threshold to 100% of the nodes' contributions by default. " +
+                    "To instead wait for a single contribution, set thresholdType = 'absolute'"
+                )
+                this.#thresholdType = 'relative'
+            } else {
+                this.#thresholdType = thresholdType
+            }
+        }
+    }
+
+    /** Checks whether the contributions buffer is full. */
+    override isFull(): boolean {
+        // Make sure that we are over the minimum number of participants
+        // if specified
+        if (this.#minNbOfParticipants !== undefined &&
+            this.nodes.size < this.#minNbOfParticipants) return false
+
+        const thresholdValue =
+            this.#thresholdType == 'relative'
+                ? this.#threshold * this.nodes.size
+                : this.#threshold;
+
+        return (this.contributions.get(0)?.size ?? 0) >= thresholdValue;
+    }
+
+    set minNbOfParticipants(minNbOfParticipants: number) {
+        this.#minNbOfParticipants = minNbOfParticipants
+    }
+
+    override _add(nodeId: client.NodeID, contribution: WeightsContainer): void {
+
+        this.log(
+            this.contributions.hasIn([0, nodeId]) ? AggregationStep.UPDATE : AggregationStep.ADD,
+            nodeId,
+        );
+
+        const beta: number = 0.9; // Momentum parameter
+
+        // Apply worker-side momentum
+
+        const prevMomentum: WeightsContainer = this.contributions.getIn([0, nodeId]) as WeightsContainer;
+        let momentum: WeightsContainer;
+
+        if (prevMomentum) {
+            // m_t = (1 - beta) * grad + beta * m_{t-1}
+            momentum = contribution.map((g, i) =>
+                g.mul(1 - beta).add(prevMomentum.weights[i].mul(beta))
+            );
+        } else {
+            momentum = contribution.map(g => g.mul(1 - beta));
+        }
+
+        this.contributions = this.contributions.setIn([0, nodeId], momentum);
+    }
+
+    override aggregate(): WeightsContainer {
+        const currentContributions = this.contributions.get(0);
+
+        if (currentContributions === undefined)
+            throw new Error("aggregating without any contribution");
+
+        this.log(AggregationStep.AGGREGATE);
+
+        let v = aggregation.avg(currentContributions.values());
+
+        for (let iter = 0; iter < this.maxIterations; iter++) {
+            const updated = currentContributions.map(m => {
+                const diff = m.sub(v)
+                const norm = this.euclideanNorm(diff)
+                const scale = tf.tidy(() =>
+                    tf.minimum(tf.scalar(1), tf.div(tf.scalar(this.clippingRadius), norm))
+                )
+                return diff.mul(scale).add(v)
+            })
+
+            v = aggregation.avg(updated.values());
+        }
+
+        return v
+    }
+
+    private euclideanNorm(w: WeightsContainer): tf.Scalar {
```

Comment:
doesn't need access to `this`, better to extract it out of the class (in the same file)